File: python-backend/services/snippet_emitter.py
"""SSE snippet emitter — Phase B prep.

A small helper that yields events conforming to the SSE shape documented in
CONTRACTS.md. Phase A endpoints don't use this yet; it's wired up in Phase B
when each chat / workflow phase emits a ``code-snippet`` followed by a
``result``.

Each method is an async generator so callers can ``async for`` and forward the
strings straight into ``sse_starlette.EventSourceResponse``.
"""

from __future__ import annotations

import json
from collections.abc import AsyncIterator


def _format(event_type: str, payload: dict) -> str:
    """Turn a dict into the ``event: <type>\\ndata: <json>\\n\\n`` SSE frame."""
    return f"event: {event_type}\ndata: {json.dumps(payload)}\n\n"


class SnippetEmitter:
    """Builds CONTRACTS.md-compliant SSE events for one walkthrough stage.

    The emitter is stateless beyond the stage name and (optional) phase tracking
    bookkeeping. Each method is an async generator yielding exactly one frame so
    callers can compose streams without juggling formatting.
    """

    def __init__(self, stage: str, language: str = "python") -> None:
        self.stage = stage
        self.language = language
        self._phase_id: str | None = None
        self._phase_index: int | None = None
        self._phase_total: int | None = None

    async def phase_start(
        self,
        phase_id: str,
        phase_index: int,
        phase_total: int,
    ) -> AsyncIterator[str]:
        """Emit a ``phase-start`` frame and remember the phase context."""
        self._phase_id = phase_id
        self._phase_index = phase_index
        self._phase_total = phase_total
        yield _format(
            "phase-start",
            {
                "type": "phase-start",
                "stage": self.stage,
                "phase_id": phase_id,
                "phase_index": phase_index,
                "phase_total": phase_total,
                "status": "running",
            },
        )

    async def code(self, code: str, log: str | None = None) -> AsyncIterator[str]:
        """Emit a ``code-snippet`` frame for a chunk of teaching code."""
        payload: dict = {
            "type": "code-snippet",
            "stage": self.stage,
            "language": self.language,
            "code": code,
            "status": "pending",
        }
        if self._phase_id is not None:
            payload["phase_id"] = self._phase_id
            payload["phase_index"] = self._phase_index
            payload["phase_total"] = self._phase_total
        if log is not None:
            payload["log"] = log
        yield _format("code-snippet", payload)

    async def result(
        self,
        log: str,
        view_in_sandbox: dict | None = None,
    ) -> AsyncIterator[str]:
        """Emit a ``result`` frame describing the outcome of the previous code."""
        payload: dict = {
            "type": "result",
            "stage": self.stage,
            "status": "done",
            "log": log,
        }
        if self._phase_id is not None:
            payload["phase_id"] = self._phase_id
            payload["phase_index"] = self._phase_index
            payload["phase_total"] = self._phase_total
        if view_in_sandbox is not None:
            payload["view_in_sandbox"] = view_in_sandbox
        yield _format("result", payload)

    async def phase_end(self) -> AsyncIterator[str]:
        """Emit a ``phase-end`` frame for the current phase."""
        payload: dict = {
            "type": "phase-end",
            "stage": self.stage,
            "status": "done",
        }
        if self._phase_id is not None:
            payload["phase_id"] = self._phase_id
            payload["phase_index"] = self._phase_index
            payload["phase_total"] = self._phase_total
        yield _format("phase-end", payload)

    async def done(self) -> AsyncIterator[str]:
        """Emit the terminal ``done`` frame."""
        yield _format("done", {"type": "done"})

    async def error(self, msg: str) -> AsyncIterator[str]:
        """Emit an ``error`` frame; callers should stop the stream after."""
        yield _format(
            "error",
            {
                "type": "error",
                "stage": self.stage,
                "status": "error",
                "error": msg,
            },
        )

File: node-backend/src/services/snippetEmitter.ts
import type { Response } from "express";

/**
 * "View in Sandbox" link payload — see CONTRACTS.md SSE event shape.
 */
export interface ViewInSandbox {
  label: string;
  url: string;
}

/**
 * SSE event helper for streaming code-snippet walkthroughs.
 *
 * Phase A: not used by route handlers; the stub endpoints write SSE
 * inline. This class exists so Phase B handlers can swap straight
 * to it without rewriting plumbing.
 *
 * Each method writes exactly one SSE event of the shape defined in
 * CONTRACTS.md ("type", "stage", optional fields). Construction sets the
 * three SSE response headers; callers do not need to touch them.
 */
export class SnippetEmitter {
  private readonly res: Response;
  private readonly stage: string;

  constructor(res: Response, stage: string) {
    this.res = res;
    this.stage = stage;
    res.setHeader("Content-Type", "text/event-stream");
    res.setHeader("Cache-Control", "no-cache");
    res.setHeader("Connection", "keep-alive");
  }

  /**
   * Emits a `phase-start` event. `idx` is 1-based for human-readable display.
   */
  phaseStart(phaseId: string, idx: number, total: number): void {
    this.write("phase-start", {
      type: "phase-start",
      stage: this.stage,
      phase_id: phaseId,
      phase_index: idx,
      phase_total: total,
      status: "running",
    });
  }

  /**
   * Emits a `code-snippet` event. The snippet is shown verbatim in the UI;
   * an optional log string explains what the line does.
   */
  code(snippet: string, opts?: { log?: string }): void {
    const payload: Record<string, unknown> = {
      type: "code-snippet",
      stage: this.stage,
      code: snippet,
      language: "typescript",
      status: "pending",
    };
    if (opts?.log !== undefined) payload.log = opts.log;
    this.write("code-snippet", payload);
  }

  /**
   * Emits a `result` event after the snippet ran. `viewInSandbox` is the
   * deep-link the frontend renders as a "View in Sandbox" button.
   */
  result(opts: { log: string; viewInSandbox?: ViewInSandbox }): void {
    const payload: Record<string, unknown> = {
      type: "result",
      stage: this.stage,
      status: "done",
      log: opts.log,
    };
    if (opts.viewInSandbox) payload.view_in_sandbox = opts.viewInSandbox;
    this.write("result", payload);
  }

  /**
   * Emits a `phase-end` event. Pairs with the most recent `phaseStart`.
   */
  phaseEnd(): void {
    this.write("phase-end", {
      type: "phase-end",
      stage: this.stage,
      status: "done",
    });
  }

  /**
   * Emits a terminal `done` event and closes the stream.
   */
  done(): void {
    this.write("done", { type: "done" });
    this.res.end();
  }

  /**
   * Emits a custom SSE event with a caller-supplied payload. Used by the
   * chat handlers (manual + auto) which need to add chat-specific fields
   * like `provider`, `model`, `usage` that the generic `result()` helper
   * doesn't carry.
   */
  emit(eventType: string, payload: Record<string, unknown>): void {
    this.write(eventType, payload);
  }

  /**
   * Emits a terminal `error` event and closes the stream.
   */
  error(msg: string): void {
    this.write("error", {
      type: "error",
      stage: this.stage,
      status: "error",
      error: msg,
    });
    this.res.end();
  }

  /**
   * Internal: write one SSE event line per the spec
   *   `event: <name>\ndata: <json>\n\n`.
   */
  private write(eventType: string, payload: unknown): void {
    this.res.write(
      `event: ${eventType}\ndata: ${JSON.stringify(payload)}\n\n`
    );
  }
}
